web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
+ */
+export function useKeyBasedPagination<T>({
+  fetchFunc,
+  clusterId,
+  filter,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const abortController = useRef<AbortController | null>(null);
+  const { attempt, setAttempt } = useAttempt();
+  const [finished, setFinished] = useState(false);
+  const [resources, setResources] = useState<T[]>([]);
+  const [startKey, setStartKey] = useState(null);
+
+  useEffect(() => {
+    abortController.current?.abort();
+    abortController.current = null;
+
+    setAttempt({ status: '', statusText: '' });
+    setFinished(false);
+    setResources([]);
+    setStartKey(null);
+  }, [clusterId, ...resourceFilterToHookDeps(filter)]);
+
+  // This is an additional countermeasure to prevent fetching if `fetch` is
+  // called multiple times per React state reconciliation cycle.
+  let fetchCalled = false;
+  const fetch = async (force: boolean) => {
+    if (
+      finished ||
+      fetchCalled ||
+      (!force &&
+        (attempt.status === 'processing' || attempt.status === 'failed'))
+    ) {
+      return;
+    }
+
+    try {
+      fetchCalled = true;
+      setAttempt({ status: 'processing' });
+      const limit = resources.length > 0 ? fetchMoreSize : initialFetchSize;
+      abortController.current?.abort();
+      abortController.current = new AbortController();
+      const res = await fetchFunc(
+        clusterId,
+        {
+          ...filter,
+          limit,
+          startKey,
+        },
+        abortController.current.signal
+      );
+      abortController.current = null;
+      setResources([...resources, ...res.agents]);
+      setStartKey(res.startKey);
+      if (!res.startKey) {
+        setFinished(true);
+      }
+      setAttempt({ status: 'success' });
+    } catch (err) {
+      // Aborting is not really an error here.
+      if (isAbortError(err)) {
+        setAttempt({ status: '', statusText: '' });
+        return;
+      }
+      setAttempt({ status: 'failed', statusText: err.message });
+    }
+  };
+
+  return {
+    fetch: () => fetch(false),
+    forceFetch: () => fetch(true),
+    attempt,
+    resources,
+    finished,
+  };
+}
+
+const isAbortError = (err: any): boolean =>
+  (err instanceof DOMException && err.name === 'AbortError') ||
+  (err.cause && isAbortError(err.cause));
```

Comment:
Can you provide a reference or some explanation for why we think this is an exhaustive check across browsers?

Reply:
Neither caniuse.com, nor [MDN](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) mention any browser-specific behaviors about the `AbortError`. Same for `Error.cause`.